File: www/canvas.js
//
//  Canvas wrapper class
// 

function get_canvas_context(element) {
    var canvas = obj(element);
    if (canvas)
	return canvas.getContext('2d');
    return null;
}
    
    /* Canvas stuff */
    function CanvasClass(o) {
	return this;
    }

    CanvasClass.prototype.save = function(canvas) {
	var ctx = get_canvas_context(canvas);
	if (ctx) ctx.save();
    }

    CanvasClass.prototype.restore = function(canvas) {
	var ctx = get_canvas_context(canvas);
	if (ctx) ctx.restore();
    }

    CanvasClass.prototype.scale = function (canvas,x,y) {
	var ctx = get_canvas_context(canvas);
	if (ctx) ctx.scale(x,y);
    }

    CanvasClass.prototype.rotate = function (canvas,a) {
	var ctx = get_canvas_context(canvas);
	if (ctx) ctx.rotate(a);
    }

    CanvasClass.prototype.translate = function (canvas,x,y) {
	var ctx = get_canvas_context(canvas);
	if (ctx) ctx.translate(x,y);
    }

    CanvasClass.prototype.transform = function (canvas,a,b,c,d,e,f) {
	var ctx = get_canvas_context(canvas);
	if (ctx) ctx.transform(a,b,c,e,d,f);
    }

    CanvasClass.prototype.setTransform = function (canvas,a,b,c,d,e,f) {
	var ctx = get_canvas_context(canvas);
	if (ctx) ctx.setTransform(a,b,c,e,d,f);
    }


    CanvasClass.prototype.globalAlpha = function(canvas,a) {
	var ctx = get_canvas_context(canvas);
	if (ctx) ctx.globalAlpha = a;
    }

    CanvasClass.prototype.globalCompositeOperation = function(canvas,op) {
	var ctx = get_canvas_context(canvas);
	if (ctx) ctx.globalCompositeOperation = op;
    }

    CanvasClass.prototype.strokeStyle = function(canvas,style) {
	var ctx = get_canvas_context(canvas);
	if (ctx) ctx.strokeStyle = style;
    }

    CanvasClass.prototype.fillStyle = function(canvas,style) {
	var ctx = get_canvas_context(canvas);
	if (ctx) ctx.fillStyle = style;
    }

    /* Fixme: add gradient stuff */

    CanvasClass.prototype.lineWidth = function(canvas,w) {
	var ctx = get_canvas_context(canvas);
	if (ctx) ctx.lineWidth = w;
    }

    CanvasClass.prototype.lineCap = function(canvas,cap) {
	var ctx = get_canvas_context(canvas);
	if (ctx) ctx.lineCap = cap;
    }

    CanvasClass.prototype.lineJoin = function(canvas,join) {
	var ctx = get_canvas_context(canvas);
	if (ctx) ctx.lineJoin = join;
    }
    
    CanvasClass.prototype.miterLimit = function(canvas,lim) {
	var ctx = get_canvas_context(canvas);
	if (ctx) ctx.miterLimit = lim;
    }

    /* Shadow */

    CanvasClass.prototype.shadowOffsetX = function(canvas,x) {
	var ctx = get_canvas_context(canvas);
	if (ctx) ctx.shadowOffsetX = x;
    }

    CanvasClass.prototype.shadowOffsetY = function(canvas,y) {
	var ctx = get_canvas_context(canvas);
	if (ctx) ctx.shadowOffsetX = Y;
    }

    CanvasClass.prototype.shadowBlur = function(canvas,b) {
	var ctx = get_canvas_context(canvas);
	if (ctx) ctx.shadowBlur = b;
    }

    CanvasClass.prototype.shadowColor = function(canvas,color) {
	var ctx = get_canvas_context(canvas);
	if (ctx) ctx.shadowColor = color;
    }

    /* rects */
    CanvasClass.prototype.clearRect = function(canvas,x,y,w,h) {
	var ctx = get_canvas_context(canvas);
	if (ctx) ctx.clearRect(x, y, w, h);
    }

    CanvasClass.prototype.fillRect = function(canvas,x,y,w,h) {
	var ctx = get_canvas_context(canvas);
	if (ctx) ctx.fillRect(x, y, w, h);
    }

    CanvasClass.prototype.strokeRect = function(canvas,x,y,w,h) {
	var ctx = get_canvas_context(canvas);
	if (ctx) ctx.strokeRect(x, y, w, h);
    }

    /* Path API */
    CanvasClass.prototype.beginPath = function(canvas) {
	var ctx = get_canvas_context(canvas);
	if (ctx) ctx.beginPath();
    }

    CanvasClass.prototype.closePath = function(canvas) {
	var ctx = get_canvas_context(canvas);
	if (ctx) ctx.closePath();
    }

    CanvasClass.prototype.moveTo = function(canvas,x,y) {
	var ctx = get_canvas_context(canvas);
	if (ctx) ctx.moveTo(x,y);
    }

    CanvasClass.prototype.lineTo = function(canvas,x,y) {
	var ctx = get_canvas_context(canvas);
	if (ctx) ctx.lineTo(x,y);
    }

    CanvasClass.prototype.quadraticCurveTo = 
    function(canvas,cp1x,cp1y,x,y) {
	var ctx = get_canvas_context(canvas);
	if (ctx) ctx.quadraticCurveTo(cp1x,cp1y,x,y);
    }

    CanvasClass.prototype.bezierCurveTo = 
    function(canvas,cp1x, cp1y, cp2x, cp2y, x, y) {
	var ctx = get_canvas_context(canvas);
	if (ctx) ctx.bezierCurveTo(cp1x,cp1y,cp2x,cp2y,x,y);
    }

    CanvasClass.prototype.arcTo = 
    function(canvas, x1,y1,x2,y2,r) {
	var ctx = get_canvas_context(canvas);
	if (ctx) ctx.arcTo(x1,y1,x2,y2,r);
    }
    
    CanvasClass.prototype.rect = function(canvas,x,y,w,h) {
	var ctx = get_canvas_context(canvas);
	if (ctx) ctx.rect(x, y, w, h);
    }

    CanvasClass.prototype.arc = function(canvas,x,y,r,a0,a1,acw) {
	var ctx = get_canvas_context(canvas);
	if (ctx) ctx.arc(x,y,r,a0,a1,acw);
    }

    CanvasClass.prototype.fill = function(canvas) {
	var ctx = get_canvas_context(canvas);
	if (ctx) ctx.fill();
    }

    CanvasClass.prototype.stroke = function(canvas) {
	var ctx = get_canvas_context(canvas);
	if (ctx) ctx.stroke();
    }

    CanvasClass.prototype.clip = function(canvas) {
	var ctx = get_canvas_context(canvas);
	if (ctx) ctx.clip();
    }

    CanvasClass.prototype.isPointInPath = function(canvas, x, y) {
	var ctx = get_canvas_context(canvas);
	if (ctx) return ctx.isPointInPath(canvas, x, y);
	else return false;
    }

    /* drawFocusRing */

    CanvasClass.prototype.font = function(canvas, f) {
	var ctx = get_canvas_context(canvas);
	if (ctx) ctx.font = f;
    }

    CanvasClass.prototype.textAlign = function(canvas, a) {    
	var ctx = get_canvas_context(canvas);
	if (ctx) ctx.textAlign = a;
    }

    CanvasClass.prototype.textBaseline = function(canvas, b) {    
	var ctx = get_canvas_context(canvas);
	if (ctx) ctx.textBaseline = b;
    }

    CanvasClass.prototype.fillText = function(canvas,text,x,y,maxWidth) {
	var ctx = get_canvas_context(canvas);
	if (ctx) ctx.fillText(text,x,y,maxWidth);
    }

    CanvasClass.prototype.strokeText = function(canvas,text,x,y,maxWidth) {
	var ctx = get_canvas_context(canvas);
	if (ctx) ctx.strokeText(text,x,y,maxWidth);
    }

    CanvasClass.prototype.measureText = function(canvas,text) {
	var ctx = get_canvas_context(canvas);
	if (ctx) return ctx.measureText(text);
	else return -1;
    }

    CanvasClass.prototype.drawImage2 = 
    function(canvas,image,x,y) {
	var ctx = get_canvas_context(canvas);
	var ni  = document.images[image];
	console.log("drawImage2: name=%s", image);
	console.log(ni);
	if (ctx) ctx.drawImage(ni,x,y);
    }

    CanvasClass.prototype.drawImage4 = 
    function(canvas,image,x,y,w,h) {
	var ctx = get_canvas_context(canvas);
	var ni  = document.images[image];
	console.log("drawImage: name=%s", image);
	console.log(ni);
	if (ctx) ctx.drawImage(ni,x,y,w,h);
    }

    CanvasClass.prototype.drawImage8 = 
    function(canvas,image,x1,y1,w1,h1,x2,y2,w2,h2) {
	var ctx = get_canvas_context(canvas);
	var ni  = document.images[image];
	console.log("drawImage: name=%s", image);
	console.log(ni);
	if (ctx) ctx.drawImage(ni,x1,y1,w1,h1,x2,y2,w2,h2);
    }

    function do_scroll(ctx, x,y,w,h,dx,dy,clearStyle) {
	var ax = (dx >= 0) ? dx : -dx; // abs(dx)
	var ay = (dy >= 0) ? dy : -dy; // abs(dy)
	var sx = x + ((dx >= 0) ? 0  : ax);
	var sy = y + ((dy >= 0) ? 0  : ay);
	var tx = x + ((dx > 0) ? ax : 0);
	var ty = y + ((dy > 0) ? ay : 0);
	ctx.drawImage(ctx.canvas,
		      sx,sy,w-ax,h-ay,
		      tx,ty,w-ax,h-ay);
	if (clearStyle) {
	    var cx = x + ((dx < 0) ? (w-ax-1) : 0);
	    var cy = y + ((dy < 0) ? (h-ay-1) : 0);
	    var cw = ax ? ax : w;
	    var ch = ay ? ay : h;
	    ctx.save();
	    ctx.fillStyle = clearStyle;
	    ctx.fillRect(cx,cy,cw,ch);
	    ctx.restore();
	}
    }

    function do_load_image(ctx,canvas,name,url) {
	var img = document.images[name];
	if (!img) {
	    img = new Image();
	    document.images[name] = img;
	    img.onload = function() {
		var func = obj('page').imageLoaded;
		console.log(obj('page'));
		console.log("func = %s\n", func);
		func.call(Bert.atom('loaded'),canvas.substr(6),name);
	    }
	    console.log("create image: name=%s", name);
	}
	img.src = url;
	console.log("image: set url=%d", url);
    }

    /* special hacks */
    CanvasClass.prototype.scroll = function(canvas,x,y,w,h,dx,dy,clearStyle) {
	var ctx = get_canvas_context(canvas);
	if (ctx) do_scroll(ctx, x,y,w,h,dx,dy,clearStyle);
    }

    CanvasClass.prototype.loadImage = function(canvas,name,url) {
	var ctx = get_canvas_context(canvas);
	if (ctx) do_load_image(ctx,canvas,name,url);
    }

    CanvasClass.prototype.batch = function(canvas, commands) {
	var ctx = get_canvas_context(canvas);

	if (!ctx)
	    return;
	for (i = 0; i < commands.length; i++) {
	    var argv = commands[i];
	    switch(argv[0]) {
	    case 'save':  ctx.save(); break;
	    case 'restore': ctx.restore(); break;
	    case 'scale': ctx.scale(argv[1], argv[2]); break;
	    case 'rotate': ctx.rotate(argv[1]); break;
	    case 'translate': ctx.translate(argv[1],argv[2]); break;
	    case 'transform':
		ctx.transform(argv[1],argv[2],argv[3],argv[4],argv[5]);
		break;
	    case 'setTransform':
		ctx.setTransform(argv[1],argv[2],argv[3],argv[4],argv[5]);
		break;
	    case 'globalAlpha':	ctx.globalAlpha=argv[1]; break;
	    case 'globalCopositeOperation':
		ctx.globalCopositeOperation=argv[1]; break;
	    case 'strokeStyle': ctx.strokeStyle=argv[1]; break;
	    case 'fillStyle':   ctx.fillStyle=argv[1]; break;
	    case 'lineWidth':   ctx.lineWidth=argv[1]; break;
	    case 'lineCap':     ctx.lineCap=argv[1]; break;
	    case 'lineJoin':    ctx.lineJoin=argv[1]; break;
	    case 'miterLimit':  ctx.miterLimit=argv[1]; break;
	    case 'shadowOffsetX': ctx.shadowOffsetX=argv[1]; break;
	    case 'shadowOffsetY': ctx.shadowOffsetY=argv[1]; break;
	    case 'shadowBlur':    ctx.shadowBlur=argv[1]; break;
	    case 'shadowColor':   ctx.shadowColor=argv[1]; break;
	    case 'clearRect': 
		ctx.clearRect(argv[1],argv[2],argv[3],argv[4]);
		break;
	    case 'fillRect': 
		ctx.fillRect(argv[1],argv[2],argv[3],argv[4]);
		break;
	    case 'strokeRect': 
		ctx.strokeRect(argv[1],argv[2],argv[3],argv[4]);
		break;
	    case 'beginPath': ctx.beginPath(); break;
	    case 'closePath': ctx.closePath(); break;
	    case 'moveTo': ctx.moveTo(argv[1],argv[2]); break;
	    case 'lineTo': ctx.lineTo(argv[1],argv[2]); break;
	    case 'quadraticCurveTo':
		ctx.quadraticCurveTo(argv[1],argv[2],argv[3],argv[4]);
		break;
	    case 'bezierCurveTo':
		ctx.bezierCurveTo(argv[1],argv[2],argv[3],argv[4],
				  argv[5],argv[6]);
		break;
	    case 'arcTo':
		ctx.arcTo(argv[1],argv[2],argv[3],argv[4],argv[5]);
		break;
	    case 'rect':
		ctx.rect(argv[1],argv[2],argv[3],argv[4]);
		break;
	    case 'arc':
		ctx.arc(argv[1],argv[2],argv[3],argv[4],
			argv[5],argv[6]);
		break;
	    case 'fill':   ctx.fill();  break;
	    case 'stroke': ctx.stroke(); break;
	    case 'clip':   ctx.clip(); break;
	    case 'isPointInPath':
		// fixme handle return values!
		ctx.isPointInPath(argv[1],argv[2]);
		break;
	    case 'font': ctx.font = argv[1]; break;
	    case 'textAlign': ctx.textAlign = argv[1]; break;
	    case 'textBaseline': ctx.textBaseline = argv[1]; break;
	    case 'fillText':
		ctx.fillText(argv[1],argv[2],argv[3],argv[4]);
		break;
	    case 'strokeText':
		ctx.strokeText(argv[1],argv[2],argv[3],argv[4]);
		break;
	    case 'measureText':
		// fixme handle return value
		ctx.measureText(argv[1]);
		break;
	    case 'drawImage':
		switch (argv.length) {
		case 4: 
		    ctx.drawImage(argv[1],argv[2],argv[3]);
		    break;
		case 6:
		    ctx.drawImage(argv[1],argv[2],argv[3],argv[4],argv[5]);
		    break;
		case 10:
		    ctx.drawImage(argv[1],argv[2],argv[3],argv[4],argv[5],
				  argv[6],argv[7],argv[8],argv[9]);
		    break;
		}
		break;

	    case 'scroll':
		do_scroll(ctx, argv[1],argv[2],argv[3],argv[4],
			  argv[5],argv[6],argv[7]);
		break;

	    case 'loadImage':
		do_load_image(ctx,canvas,argv[1],argv[2]);
		break;
	    }
	}
    }


    var Canvas = new CanvasClass();
    

File: src/canvas.erl
%%% @author Tony Rogvall <tony@rogvall.se>
%%% @copyright (C) 2010, Tony Rogvall
%%% @doc
%%%    Canvas actions for nitrogen
%%% @end
%%% Created :  5 Sep 2010 by Tony Rogvall <tony@rogvall.se>

-module(canvas).

-include("../include/wf_canvas.hrl").

-compile(export_all).


%% state
save(Canvas) ->
    command(Canvas, save).

restore(Canvas) ->
    command(Canvas, restore).

%% Extension: Evaluate Fun under a save/restore pair
withState(Canvas, Fun) ->
    save(Canvas),
    try Fun() of
	Result ->
	    Result
    after
	restore(Canvas)
    end.

%% Extension:  beginPath ... fill
pathFill(Canvas, Fun) ->
    beginPath(Canvas),
    Fun(),
    fill(Canvas).

%% Extension:  beginPath ... stroke
pathStroke(Canvas, Fun) ->
    beginPath(Canvas),
    Fun(),
    stroke(Canvas).
    


%% transformation
scale(Canvas, X, Y) when is_number(X), is_number(Y) -> 
    command(Canvas, {scale,X,Y}).
rotate(Canvas, Angle) when is_number(Angle) -> 
    command(Canvas, {rotate, Angle}).
translate(Canvas, X, Y) when is_number(X), is_number(Y) ->
    command(Canvas, {translate,X,Y}).

transform(Canvas, {A, B, C, D, E, F})
  when is_number(A), is_number(B), is_number(C),
       is_number(D), is_number(E), is_number(F) ->    
    command(Canvas, {transform,A,B,C,D,E,F}).
transform(Canvas, A, B, C, D, E, F) 
  when is_number(A), is_number(B), is_number(C),
       is_number(D), is_number(E), is_number(F) ->
    command(Canvas, {transform,A,B,C,D,E,F}).

setTransform(Canvas, A, B, C, D, E, F)
  when is_number(A), is_number(B), is_number(C),
       is_number(D), is_number(E), is_number(F) ->
    command(Canvas, {setTransform,A,B,C,D,E,F}).

setTransform(Canvas, {A, B, C, D, E, F})
  when is_number(A), is_number(B), is_number(C),
       is_number(D), is_number(E), is_number(F) ->
    command(Canvas, {setTransform,A,B,C,D,E,F}).

%% compositing
globalAlpha(Canvas, Alpha) when is_number(Alpha) ->
    command(Canvas, {globalAlpha, Alpha}).
globalCompositeOperation(Canvas, Operation) when is_list(Operation) ->
    command(Canvas, {globalCompositeOperation, Operation}).

%% colors and styles

strokeStyle(Canvas, Style) when is_list(Style); is_atom(Style) ->
    command(Canvas, {strokeStyle,Style}).

fillStyle(Canvas, Style) when is_list(Style); is_atom(Style) ->
    command(Canvas, {fillStyle,Style}).

%% fixme gradiant/pattern
lineWidth(Canvas, Width) when is_number(Width) ->
    command(Canvas, {lineWidth, Width}).

%% Cap = butt | round | square
lineCap(Canvas, Cap) when is_atom(Cap)  ->
    command(Canvas, {lineCap, Cap}).

%% Join = round | bevel | miter
lineJoin(Canvas, Join) when is_atom(Join) ->
    command(Canvas, {lineJoin, Join}).

miterLimit(Canvas, Limit) when is_number(Limit) ->
    command(Canvas, {miterLimit, Limit}).

%% shadows
shadowOffsetX(Canvas,X) when is_number(X) ->
    command(Canvas, {shadowOffsetX, X}).

shadowOffsetY(Canvas,Y) when is_number(Y) ->
    command(Canvas, {shadowOffsetY, Y}).

shadowBlur(Canvas, Blur) when is_number(Blur) ->
    command(Canvas, {shadowBlur, Blur}).

shadowColor(Canvas,Color) when is_list(Color); is_atom(Color) -> 
    command(Canvas, {shadowColor,Color}).

%% rects    
clearRect(Canvas, {X, Y, W, H}) ->
    clearRect(Canvas, X, Y, W, H).
clearRect(Canvas, X, Y, W, H)
  when is_number(X), is_number(Y), is_number(W), is_number(H) ->
    command(Canvas, {clearRect,X,Y,W,H}).
fillRect(Canvas, {X, Y, W, H}) ->
    fillRect(Canvas, X, Y, W, H).
fillRect(Canvas, X, Y, W, H)
  when is_number(X), is_number(Y), is_number(W), is_number(H) ->
    command(Canvas, {fillRect,X,Y,W,H}).
strokeRect(Canvas, {X, Y, W, H}) ->
    strokeRect(Canvas, X, Y, W, H).
strokeRect(Canvas, X, Y, W, H) 
  when is_number(X), is_number(Y), is_number(W), is_number(H) ->
    command(Canvas, {strokeRect,X,Y,W,H}).

%% path API
beginPath(Canvas) ->
    command(Canvas, beginPath).

closePath(Canvas)  ->
    command(Canvas, closePath).

moveTo(Canvas,{X,Y})  ->
    moveTo(Canvas,X,Y).
moveTo(Canvas,X,Y) 
  when is_number(X), is_number(Y) ->
    command(Canvas, {moveTo,X,Y}).

lineTo(Canvas,{X,Y}) ->
    lineTo(Canvas, X, Y).
lineTo(Canvas,X,Y)  
  when is_number(X), is_number(Y) -> 
    command(Canvas, {lineTo,X,Y}).

quadraticCurveTo(Canvas,CX1,CY1,X,Y)
    when is_number(CX1), is_number(CY1), is_number(X), is_number(Y) -> 
    command(Canvas, {quadraticCurveTo,CX1,CY1,X,Y}).

bezierCurveTo(Canvas,CX1,CY1,CX2,CY2,X,Y)
    when is_number(CX1), is_number(CY1), is_number(CX2), is_number(CY2),
	 is_number(X), is_number(Y) -> 
    command(Canvas, {bezierCurveTo,CX1,CY1,CX2,CY2,X,Y}).

arcTo(Canvas,X1,Y1,X2,Y2,R)
  when is_number(X1), is_number(Y1), is_number(X2), is_number(Y2),
       is_number(R) ->
    command(Canvas, {arcTo,X1,Y1,X2,Y2,R}).

rect(Canvas, {X, Y, W, H}) ->
    rect(Canvas, X, Y, W, H).
rect(Canvas, X, Y, W, H)
  when is_number(X), is_number(Y), 
       is_number(W), is_number(H) ->
    command(Canvas, {rect,X,Y,W,H}).

arc(Canvas,X,Y,R,A0,A1,ACW) 
  when is_number(X), is_number(Y), is_number(R), 
       is_number(A0), is_number(A1), is_boolean(ACW) ->
    command(Canvas, {arc,X,Y,R,A0,A1,ACW}).

fill(Canvas) ->
    command(Canvas, fill).

stroke(Canvas) ->
    command(Canvas, stroke).

clip(Canvas) -> 
    command(Canvas, clip).

%% bool isPointInPath(X,Y) 
%%  boolean drawFocusRing(in Element element, in double xCaret, in double yCaret, in optional boolean canDrawCustom)
font(Canvas,Font) 
  when is_list(Font); is_atom(Font)  -> 
    command(Canvas, {font, Font}).

textAlign(Canvas,Align)
  when is_atom(Align) ->
    command(Canvas, {textAlign,Align}).

textBaseline(Canvas,Base) 
  when is_atom(Base) -> 
    command(Canvas, {textBaseline,Base}).

fillText(Canvas, Text, X, Y) 
  when is_list(Text), is_number(X), is_number(Y) ->
    command(Canvas,{fillText,Text,X,Y,null}).

fillText(Canvas, Text, X, Y, MaxWidth) 
  when is_list(Text), is_number(X), is_number(Y), is_number(MaxWidth) ->
    command(Canvas,{fillText,Text,X,Y,MaxWidth}).

strokeText(Canvas, Text, X, Y)
  when is_list(Text), is_number(X), is_number(Y) ->
    command(Canvas,{strokeText,Text,X,Y,null}).

strokeText(Canvas, Text, X, Y, MaxWidth) 
  when is_list(Text), is_number(X), is_number(Y), is_number(MaxWidth) ->
    command(Canvas,{strokeText,Text,X,Y,MaxWidth}).

    
%% TextMetrics mesureText(Text)
%%  void drawImage(in HTMLImageElement image, in double dx, in double dy, in optional double dw, in double dh);
drawImage(Canvas, Image, Dx, Dy)
  when is_list(Image), is_number(Dx), is_number(Dy) ->
    command(Canvas,{drawImage2,Image,Dx,Dy}).

drawImage(Canvas, Image, Dx, Dy, Dw, Dh) 
  when is_list(Image), is_number(Dx), is_number(Dy), 
       is_number(Dw), is_number(Dh) ->
    command(Canvas,{drawImage4,Image,Dx,Dy,Dw,Dh}).

drawImage(Canvas, Image, Sx, Sy, Sw, Sh, Dx, Dy, Dw, Dh) 
  when is_list(Image), 
       is_number(Sx), is_number(Sy), 
       is_number(Sw), is_number(Sh),
       is_number(Dx), is_number(Dy), 
       is_number(Dw), is_number(Dh) ->
    command(Canvas,{drawImage8,Image,Sx,Sy,Sw,Sh,Dx,Dy,Dw,Dh}).

%%/ pixel manipulation
%%  ImageData createImageData(in double sw, in double sh);
%%  ImageData createImageData(in ImageData imagedata);
%%  ImageData getImageData(in double sx, in double sy, in double sw, in double sh);
%%  void putImageData(in ImageData imagedata, in double dx, in double dy, in optional double dirtyX, in double dirtyY, in double dirtyWidth, in double dirtyHeight);

%% Image hacks
scroll(Canvas, X, Y, W, H, Dx, Dy)
  when is_number(X), is_number(Y), is_number(W), is_number(H),
       is_number(Dx), is_number(Dy) ->
    command(Canvas,{scroll,X,Y,W,H,Dx,Dy}).

scroll(Canvas, X, Y, W, H, Dx, Dy, Clear)
  when is_number(X), is_number(Y), is_number(W), is_number(H),
       is_number(Dx), is_number(Dy), is_list(Clear) ->
    command(Canvas,{scroll,X,Y,W,H,Dx,Dy,Clear}).

batch(Canvas, Commands) when is_list(Commands) ->
    wf:wire(Canvas, #canvas_action { command={batch,Commands} }). 

command(Canvas, Command) ->
    wf:wire(Canvas, #canvas_action { command=Command }).

%%
%% Loading image
%%
loadImage(Canvas, Name, Url) ->
    command(Canvas,{loadImage,Name,Url}).
